File: quick_sort.py
import math

def quick_sort(arr):
  
  if (len(arr)<2):
    return arr
  
  i=math.floor(len(arr)/2)
  pivot=arr[i]
  left=[]
  middle=[]
  right=[]
  for x in arr:
    if(x<pivot):
      left.append(x)
    elif(x==pivot):
      middle.append(x)
    else:
      right.append(x)
  
  return quick_sort(left)+middle+quick_sort(right)


  
arr=[7,1,20,60,4,5,22,3]
arr=quick_sort(arr)
for x in arr:
  print(x)

File: quickSort.c
#include <stdio.h>

//Unlike merge in mergeSort, the partition procedure in quickSort is in place
int partition(int *arr, int start, int end){
  //pick a pivot value and put it at the end of the (sub)array
  int pivot_value=arr[(start+end)/2];
  arr[(start+end)/2]=arr[end];
  arr[end]=pivot_value;
  
  int i=start-1;
  int j;
  
  for(j=start;j<end;j++){
    if(arr[j]<pivot_value){
      i++;
      int temp=arr[i];
      arr[i]=arr[j];
      arr[j]=temp;
    }
  }
  i++;
  arr[end]=arr[i];
  arr[i]=pivot_value;
  return i;
  
}


void quickSort(int *arr, int start, int end){
  
  if(start>=end)
    return;
  
  int pivot_index = partition(arr, start, end);
  quickSort(arr,start,pivot_index-1);
  quickSort(arr,pivot_index+1,end);
}

int main(void) {
  int arr[]={2,5,1,8,4,6,3}; //7 elements
  quickSort(arr,0,6);
  int i;
  for (i=0;i<7;i++)
    printf("%d ", arr[i]);
  printf("\n");
  return 0;
}
